### PR TITLE
fix for Jira ML-220

### PR DESCRIPTION
**Pull request by fazlan-nazeem**

### CODE REVIEW

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java [hunk 1]
```diff
@@ -308,11 +308,20 @@ public void insertHyperParameters(long analysisId, List<MLHyperParameter> hyperP
     public boolean isValidModelId(int tenantId, String userName, long modelId) throws DatabaseHandlerException;
 
     /**
+     * Check whether the model is in a valid state
+     * @param modelId
```
Comment (cdathuraliya) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java:312]: Add missing comments.

Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java:312]: This was fixed by a later commit

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java [hunk 2]
```diff
@@ -2637,4 +2637,35 @@ public void updateSamplePoints(long datasetVersionId, SamplePoints samplePoints)
...
+        ResultSet result = null;
+        PreparedStatement statement = null;
+        try {
+            connection = dbh.getDataSource().getConnection();
+            statement = connection.prepareStatement(SQLQueries.GET_MODEL_STATUS);
+            statement.setLong(1, modelId);
+            statement.setInt(2, tenantId);
+            statement.setString(3, userName);
+            result = statement.executeQuery();
+            if (result.first()) {
```
Comment (upul) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java:2653]: It looks like we can replace two IFs by a  AND construct.
Also, I think it is possible to remove ELSE construct.

Could you please check this?

Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java:2653]: if we combine it with an AND construct , the exception will be thrown when the model is found but the status in not complete. we need to throw the exception only when the model is not found.

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java [hunk 3]
```diff
@@ -308,11 +308,20 @@ public void insertHyperParameters(long analysisId, List<MLHyperParameter> hyperP
...
+     */
+    public boolean isValidModelStatus(long modelId, int tenantId, String userName) throws DatabaseHandlerException;
+    
+    /**
      * Get the id of the dataset-version used to generate the model
      * @param modelId  unique id of the model
      * @return
      * @throws DatabaseHandlerException
      */
+
```
Comment (lafernando) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java:324]: Remove unnecessary whitespace.

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java [hunk 4]
```diff
@@ -308,11 +308,20 @@ public void insertHyperParameters(long analysisId, List<MLHyperParameter> hyperP
...
     /**
+     * Check whether the model is in a valid state
+     * @param modelId
+     * @return
+     * @throws DatabaseHandlerException
+     */
+    public boolean isValidModelStatus(long modelId, int tenantId, String userName) throws DatabaseHandlerException;
+    
+    /**
      * Get the id of the dataset-version used to generate the model
```
Comment (lafernando) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java:319]: End the comment with a full stop. Fix in all the places.

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java [hunk 5]
```diff
@@ -2637,4 +2637,35 @@ public void updateSamplePoints(long datasetVersionId, SamplePoints samplePoints)
...
+                if (result.getString(1).equalsIgnoreCase(MLConstants.MODEL_STATUS_COMPLETE)) {
```
Comment (prabathabey) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java:2654]: As a good practice, always compare values retrieved from other software components/systems with known constants to avoid redundant null checks.

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java [hunk 6]
```diff
@@ -2637,4 +2637,35 @@ public void updateSamplePoints(long datasetVersionId, SamplePoints samplePoints)
...
+                }
+            } else {
+                throw new DatabaseHandlerException("Failed to find the model for model id " + modelId);
+            }
+        } catch (SQLException e) {
+            throw new DatabaseHandlerException(
+                    "An error has occurred while fetching the status of the model for model id: " + modelId + ": "
+                            + e.getMessage(), e);
+        } finally {
+            // Close the database resources.
```
Comment (prabathabey) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java:2665]: Redundant comment? The method call itself is self-explanatory IMO.

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java [hunk 7]
```diff
@@ -2637,4 +2637,35 @@ public void updateSamplePoints(long datasetVersionId, SamplePoints samplePoints)
...
+            MLDatabaseUtils.closeDatabaseResources(connection, statement, result);
+        }
+        // consider anything other than "Complete" status as an invalid model.
```
Comment (prabathabey) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java:2668]: Use proper linguistics for comments. Use capital letters to begin the sentences with.

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java (hunk 3 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java:324]: Fixed with PR https://github.com/wso2/carbon-ml/pull/201

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java (hunk 5 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java:2654]: fixed with PR https://github.com/wso2/carbon-ml/pull/201

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java (hunk 6 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java:2665]: Removed comment with PR https://github.com/wso2/carbon-ml/pull/201

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java (hunk 7 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/internal/MLDatabaseService.java:2668]: Comment started with capital letter. Fixed with PR  https://github.com/wso2/carbon-ml/pull/201

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java [hunk 8]
```diff
@@ -391,6 +400,13 @@ public String streamingPredict(int tenantId, String userName, long modelId, Stri
             throw new MLModelHandlerException(msg);
         }
 
+        if (!isValidModelStatus(modelId, tenantId, userName)) {
+            String msg = String
+                    .format("This model cannot be used for prediction. Status of the model for model id: %s for tenant: %s and user: %s is not 'Complete'",
```
Comment (malakasilva) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java:405]: Better have the string as a constant

Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java:405]: we are using "msg" as a string constant. The first argument string for format function has format specifiers in it. Therefore it was decided to use an inline string as the first argument just as in other places of the code.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java [hunk 9]
```diff
@@ -158,6 +158,15 @@ public boolean isValidModelId(int tenantId, String userName, long modelId) throw
             throw new MLModelHandlerException(e.getMessage(), e);
         }
     }
+    
+    public boolean isValidModelStatus(long modelId, int tenantId, String userName) throws MLModelHandlerException {
+        try {
+            return databaseService.isValidModelStatus(modelId, tenantId, userName);
+        } catch (DatabaseHandlerException e) {
+            throw new MLModelHandlerException(e.getMessage(), e);
```
Comment (sinthuja) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java:166]: Rather than simply passing the error message, add some context on the error message regarding the intended operation.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java [hunk 10]
```diff
@@ -158,6 +158,15 @@ public boolean isValidModelId(int tenantId, String userName, long modelId) throw
             throw new MLModelHandlerException(e.getMessage(), e);
         }
     }
+
```
Comment (Kishanthan) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java:161]: As a best practice, always add javadocs comments to public methods.

Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java:161]: This was fixed by an earlier PR https://github.com/wso2/carbon-ml/pull/200

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java (hunk 9 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java:166]: Added a context message. Fixed with PR https://github.com/wso2/carbon-ml/pull/202

File: components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java (hunk 4 above)
Comment (fazlan-nazeem) [inline components/ml/org.wso2.carbon.ml.database/src/main/java/org/wso2/carbon/ml/database/DatabaseService.java:319]: fixed with PR https://github.com/wso2/carbon-ml/pull/201
